great_expectations/zep/sources.py

```diff
@@ -0,0 +1,125 @@
+    or `DataAsset` types and a simplified name for those types.
+    """
+
+    type_lookup = TypeLookup()
+    __source_factories: Dict[str, SourceFactoryFn] = {}
+
+    def __init__(self, data_context: Union[DataContext, GXDataContext]):
+        self._data_context = data_context
+
+    @classmethod
+    def register_factory(
+        cls,
+        ds_type: type,
+        fn: SourceFactoryFn,
+        asset_types: List[Type[DataAsset]],
+    ) -> None:
+        """
+        Add/Register a datasource factory function.
+
+        Derives a SIMPLIFIED_NAME from the provided `Datasource` type.
+        Attaches a method called `add_<SIMPLIFIED_NAME>()`.
+
+        Also registers related `DataAsset` types.
+
+        Example
+        -------
+        `class PandasDatasource` -> `add_pandas()`
+        """
+        simplified_name = _get_simplified_name_from_type(
+            ds_type, suffix_to_remove="_datasource"
+        )
+
+        method_name = f"add_{simplified_name}"
+        LOGGER.info(
+            f"2a. Registering {ds_type.__name__} as {simplified_name} with {method_name}() factory"
+        )
+
+        pre_existing = cls.__source_factories.get(method_name)
+        if pre_existing:
+            raise ValueError(f"{simplified_name} factory already exists")
+
+        # TODO: simplify or extract the following datasource & asset type registration logic
+        asset_types = asset_types or []
+        asset_type_names = [
+            _get_simplified_name_from_type(t, suffix_to_remove="_asset")
+            for t in asset_types
+        ]
+
+        # TODO: We should namespace the asset type to the datasource so different datasources can reuse asset types.
+        already_registered_assets = set(asset_type_names).intersection(
+            cls.type_lookup.keys()
+        )
+        if already_registered_assets:
+            raise ValueError(
+                f"The following names already have a registered type - {already_registered_assets} "
+            )
+
+        for type_, name in zip(asset_types, asset_type_names):
+            cls.type_lookup[type_] = name
+            LOGGER.debug(f"'{name}' added to `type_lookup`")
+
+        cls.type_lookup[ds_type] = simplified_name
+        LOGGER.debug(f"'{simplified_name}' added to `type_lookup`")
+        cls.__source_factories[method_name] = fn  # type: ignore[assignment]
+
+    @property
+    def factories(self) -> List[str]:
+        return list(self.__source_factories.keys())
+
+    def __getattr__(self, name):
+        try:
+            fn = self.__source_factories[name]
+
+            def wrapped(*args, **kwargs):
+                datasource = fn(*args, **kwargs)
+                # TODO (bdirks): _attach_datasource_to_context to the AbstractDataContext class
+                self._data_context._attach_datasource_to_context(datasource)
+                return datasource
+
+            return wrapped
+        except KeyError:
+            raise AttributeError(name)
```

Comment:
Maybe print out the allowed keys here. Something like:
`raise AttributeError(f"No factory {name} in {self.factories}")`